File: src/main/java/shane/leetcode/problems/easy/Q1002FindCommonCharacters.java
package shane.leetcode.problems.easy;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Runtime 3 ms Beats 88.78% of users with Java
 */
public class Q1002FindCommonCharacters {

    @Test
    public void test() {
        assertThat(commonChars(new String[]{"bella", "label", "roller"})).containsExactlyInAnyOrder("e", "l", "l");
        assertThat(commonChars(new String[]{"cool", "lock", "cook"})).containsExactlyInAnyOrder("c", "o");
    }

    public List<String> commonChars(String[] words) {
        int[] arr = new int[26];
        Arrays.fill(arr, Integer.MAX_VALUE);

        for (String word : words) {
            int[] cnt = new int[26];
            for (char c : word.toCharArray()) {
                cnt[c - 'a']++;
            }
            for (int i = 0; i < 26; i++) {
                arr[i] = Math.min(arr[i], cnt[i]);
            }
        }
        List<String> answer = new ArrayList<>();
        for (int i = 0; i < 26; i++) {
            char c = (char) ('a' + i);
            for (int j = 0; j < arr[i]; j++) {
                answer.add(String.valueOf(c));
            }
        }
        return answer;
    }

}

File: src/main/kotlin/shane/leetcode/problems/easy/Q1002FindCommonCharacters.kt
package shane.leetcode.problems.easy

import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test

/**
 * Runtime: 421 ms, faster than 36.00% of Kotlin online submissions for Find Common Characters.
Memory Usage: 39.4 MB, less than 48.00% of Kotlin online submissions for Find Common Characters.

 */
class Q1002FindCommonCharacters {

    @Test
    fun test() {
        Assertions.assertThat(
            commonChars(
                arrayOf(
                    "acabcddd",
                    "bcbdbcbd",
                    "baddbadb",
                    "cbdddcac",
                    "aacbcccd",
                    "ccccddda",
                    "cababaab",
                    "addcaccd"
                )
            )
        ).isEmpty()
        Assertions.assertThat(commonChars(arrayOf("bella", "label", "roller"))).containsExactlyInAnyOrder("e", "l", "l")
        Assertions.assertThat(commonChars(arrayOf("cool", "lock", "cook"))).containsExactlyInAnyOrder("c", "o")
    }

    fun commonChars(words: Array<String>): List<String> {
        val arr = Array(words.size) { IntArray(26) }
        for (i in words.indices) {
            for (c in words[i]) {
                arr[i][c - 'a']++
            }
        }

        val answer = mutableListOf<String>()
        for (i in 0..25) {
            var min = 100
            for (j in words.indices) {
                min = minOf(min, arr[j][i])
            }
            for (j in 0 until min) {
                answer.add(('a' + i).toString())
            }
        }
        return answer
    }

}
